text/3573-is.md

```diff
@@ -0,0 +1,299 @@
+the `if` condition provides a natural extension that simplifies what would
+otherwise require complex nested conditionals. As the `let`-chains RFC notes,
+this is a feature people already expect to work.
+
+The `is` operator similarly allows developers to chain multiple match-and-bind
+operations and simplify what would otherwise require complex nested
+conditionals. However, the `is` operator allows writing and reading a pattern
+match from left-to-right, which reads more naturally in many circumstances. For
+instance, consider an expression like `x is Some(y) && y > 5`; that boolean
+expression reads more naturally from left-to-right than
+`let Some(y) = x && y > 5`.
+
+This is even more true at the end of a longer expression chain, such as
+`x.method()?.another_method().await? is Some(y)`. Rust method chaining and `?`
+and `.await` all encourage writing code that reads in operation order from left
+to right, and `is` fits naturally at the end of such a sequence.
+
+Having an `is` operator would also help to reduce the proliferation of methods
+on types such as `Option` and `Result`, by allowing prospective users of those
+methods to write a condition using `is` instead. While any such condition could
+equivalently be expressed using `let`-chains, the binding would then move
+further away from the condition expression referencing the binding, which would
+result in a less natural reading order for the expression.
+
+Consider the following examples:
+
+```rust
+if expr_producing_option().is_some_and(|v| condition(v))
+
+if let Some(v) = expr_producing_option() && condition(v)
+
+if expr_producing_option() is Some(v) && condition(v)
+```
+
+The condition using `is` is a natural translation from the `is_some_and`
+method, whereas the if-let construction requires reversing the binding of `v`
+and the expression producing the option. This seems sufficiently cumbersome in
+some cases that the absence of `is` would motivate continued use and
+development of helper methods.
+
+# Guide-level explanation
+
+Rust provides an `is` operator, which can be used in any expression:
+`EXPR is PATTERN`
+
+This operator tests if the value of `EXPR` matches the specified `PATTERN`; see
+<https://doc.rust-lang.org/reference/patterns.html> for details on patterns.
+
+If the `EXPR` matches the `PATTERN`, the `is` expression evaluates to `true`,
+and additionally binds any bindings specified in `PATTERN` in the current scope
+for code subsequently executed along the path where the `is` expression
+is known to be `true`.
+
+For example:
+
+```rust
+if an_option is Some(x) && x > 3 {
+    println!("{x}");
+}
+```
+
+The bindings in the pattern are not bound along any code path potentially
+reachable where the expression did not match:
+
+```rust
+if (an_option is Some(x) && x > 3) || (more_conditions /* x is not bound here*/) {
+    // x is not bound here
+} else {
+    // x is not bound here
+}
+// x is not bound here
+```
```

Comment:
```rust
/*1*/ let x = 9999;
/*2*/ let y = Some(4);
/*3*/ if y is Some(x) && x > 0 {
/*4*/     println!("x1 = {x}"); // x1 = 4
/*5*/ }
/*6*/ if y is Some(x) && x > 0 || cheat_code_enabled() {
/*7*/     println!("x2 = {x}"); // x2 = 9999  ⁉️
/*8*/ }
```

Just by adding that `||` on line 6 causes the `x` on line 7 refer to the outer variable on line 1 instead is surprising. 

For `if let` chains this kind of miss is impossible (and the `let` keyword makes the scope very clear), but an `is` expression can appear in any arbitrary place for expression!

I think that, even if `x` is not bound, the name `x` must still "pollute" the entire expression statement so use of `x` will generate at minimum a warn-by-default lint.